vllm/core/block_manager_v2.py

```diff
@@ -227,17 +228,89 @@ def fork(self, parent_seq: Sequence, child_seq: Sequence) -> None:
 
     def can_swap_in(self, seq_group: SequenceGroup,
                     num_lookahead_slots: int) -> bool:
-        return False
+        """
+        We go through all sequence in seq group to get their number of blocks 
+        touched and sum them up to see whether there is enough memory to swap in
+        """
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.SWAPPED):
+            block_table = self.block_tables[seq.seq_id]
+            num_touched_blocks += (
+                block_table.get_num_blocks_touched_by_append_slots(
+                    token_ids=seq.get_token_ids(),
+                    num_lookahead_slots=num_lookahead_slots,
+                ))
+        num_free_blocks = self.block_allocator.get_num_free_blocks(Device.GPU)
+        return num_free_blocks - num_touched_blocks >= self.watermark_blocks
 
     def swap_in(self, seq_group: SequenceGroup,
                 num_lookahead_slots: int) -> Dict[int, int]:
-        raise NotImplementedError
+        mapping: Dict[Block, Block] = {}
+        for seq in seq_group.get_seqs(status=SequenceStatus.SWAPPED):
+            new_block_table = BlockTable(
+                block_size=self.block_size,
+                block_allocator=self.block_allocator,
+            )
+            block_table = self.block_tables[seq.seq_id]
+
+            for cpu_block in block_table:
+                if cpu_block in mapping:
+                    gpu_block = mapping[cpu_block]
+                    self.block_allocator.increase_ref_count(
+                        Device.GPU, gpu_block.block_id())
+                else:
+                    gpu_block = new_block_table.append_by_blocks(
+                        token_ids=cpu_block.token_ids(), device=Device.GPU)
+                    mapping[cpu_block] = gpu_block
+                self.block_allocator.free(cpu_block)
+            self.block_tables[seq.seq_id] = new_block_table
+
+        # NOTE: since the memory operation in physical blocks need the
+        # relative position of CPU block to its starting address, here
+        # we need to shift the block id of cpu block back to its relative
+        # position within CPU cache.
+        block_number_mapping = {
+            cpu_block.block_id() - self.num_total_gpu_blocks:
+            gpu_block.block_id()
+            for cpu_block, gpu_block in mapping.items()
+        }
+        return block_number_mapping
 
     def can_swap_out(self, seq_group: SequenceGroup) -> bool:
-        return False
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.RUNNING):
+            num_touched_blocks += self.block_tables[
+                seq.seq_id]._num_touched_blocks()
```

Comment:
What if we replaced this with an API in the block allocator that allows users to query how many new blocks a particular sequence would consume if allocated?

Reply:
for API in the block allocator do you mean block_table? I think we need information from block table to decide how many blocks will be touched. Sure we can do similar things in `can_swap_in` with `get_num_blocks_touched_by_append_slots`, but if we have the `_num_touched_blocks` property from block table class, why don't we just leverage it?

Reply:
We can use such a `_num_touched_blocks` property, but it makes the assumption that each block has refcount 1. We want to be able to make decisions based on the _content_ of the blocks, e.g. with prefix caching. That information is stored in the lower layers (e.g. PrefixCachingBlockAllocator), so we must query that layer if we want this kind of optimization.

Reply:
Good point, I added profiling logic for the number of blocks that will be touched by swapping (different for Naive and PrefixCaching blocks). You can review it after I finish refactoring the swapping logic as well.